PoseContext.Isolate throws exceptions when used with net462

I discovered this after using Pose from a test project that's targetting `net462` and it can be reproduced by changing `Pose.Tests.csproj` to use `<TargetFramework>net462</TargetFramework>`.

I'm not sure if this is expected or not but I thought it worth mentioning in case there's a workaround or fix.

```
C:\Code\tonerdo\pose\src\Pose\Pose.csproj : warning NU1701: Package 'Mono.Reflection 1.1.0' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETStandard,Version=v2.0'.
This package may not be fully compatible with your project. [C:\Code\tonerdo\pose\test\Pose.Tests\Pose.Tests.csproj]
Build started, please wait...
C:\Code\tonerdo\pose\src\Pose\Pose.csproj : warning NU1701: Package 'Mono.Reflection 1.1.0' was restored using '.NETFramework,Version=v4.6.1' instead of the project target framework '.NETStandard,Version=v2.0'.
This package may not be fully compatible with your project.
Build completed.

Test run for C:\Code\tonerdo\pose\test\Pose.Tests\bin\Debug\net462\Pose.Tests.dll(.NETFramework,Version=v4.6.2)
Microsoft (R) Test Execution Command Line Tool Version 15.3.0-preview-20170628-02
Copyright (c) Microsoft Corporation.  All rights reserved.

Starting test execution, please wait...
Failed   Pose.Tests.StubHelperTests.TestGetShimInstance
Error Message:
 Test method Pose.Tests.StubHelperTests.TestGetShimInstance threw exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MethodAccessException: Attempt by method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<T
estGetShimInstance>b__1_1(<>c)' to access method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetShimInstance>b__1_1(<>c)' failed.
Stack Trace:
    at dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetShimInstance>b__1_1(<>c )
--- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Pose.PoseContext.Isolate(Action entryPoint, Shim[] shims)
   at Pose.Tests.StubHelperTests.TestGetShimInstance() in C:\Code\tonerdo\pose\test\Pose.Tests\Helpers\StubHelperTests.cs:line 32

Failed   Pose.Tests.StubHelperTests.TestGetShimReplacementMethod
Error Message:
 Test method Pose.Tests.StubHelperTests.TestGetShimReplacementMethod threw exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MethodAccessException: Attempt by method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<T
estGetShimReplacementMethod>b__2_1(<>c)' to access method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetShimReplacementMethod>b__2_1(<>c)' failed.
Stack Trace:
    at dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetShimReplacementMethod>b__2_1(<>c )
--- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Pose.PoseContext.Isolate(Action entryPoint, Shim[] shims)
   at Pose.Tests.StubHelperTests.TestGetShimReplacementMethod() in C:\Code\tonerdo\pose\test\Pose.Tests\Helpers\StubHelperTests.cs:line 43

Failed   Pose.Tests.StubHelperTests.TestGetMatchingShimIndex
Error Message:
 Test method Pose.Tests.StubHelperTests.TestGetMatchingShimIndex threw exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MethodAccessException: Attempt by method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<T
estGetMatchingShimIndex>b__3_2(<>c)' to access method 'DynamicClass.dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetMatchingShimIndex>b__3_2(<>c)' failed.
Stack Trace:
    at dynamic_Pose.Tests.StubHelperTests+<>c_<TestGetMatchingShimIndex>b__3_2(<>c )
--- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Pose.PoseContext.Isolate(Action entryPoint, Shim[] shims)
   at Pose.Tests.StubHelperTests.TestGetMatchingShimIndex() in C:\Code\tonerdo\pose\test\Pose.Tests\Helpers\StubHelperTests.cs:line 59

Failed   Pose.Tests.MethodRewriterTests.TestStaticMethodRewrite
Error Message:
 Test method Pose.Tests.MethodRewriterTests.TestStaticMethodRewrite threw exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.MethodAccessException: Attempt by method 'DynamicClass.stub_System.RuntimeMethodHandle_GetFuncti
onPointer(System.RuntimeMethodHandle ByRef, System.RuntimeMethodHandle, System.RuntimeTypeHandle)' to access method 'Pose.Helpers.StubHelper.GetMatchingShimIndex(System.Reflection.MethodInfo, System.Object)' fai
led.
Stack Trace:
    at stub_System.RuntimeMethodHandle_GetFunctionPointer(RuntimeMethodHandle& , RuntimeMethodHandle , RuntimeTypeHandle )
   at dynamic_Pose.Helpers.StubHelper_GetMethodPointer(MethodBase )
--- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Pose.Tests.MethodRewriterTests.TestStaticMethodRewrite() in C:\Code\tonerdo\pose\test\Pose.Tests\IL\MethodRewriterTests.cs:line 23

Failed   Pose.Tests.MethodRewriterTests.TestInstanceMethodRewrite
Error Message:
 Test method Pose.Tests.MethodRewriterTests.TestInstanceMethodRewrite threw exception:
System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. ---> System.FieldAccessException: Attempt by method 'DynamicClass.dynamic_System.Collections.Generic.List`1[
System.String]_Add(System.Collections.Generic.List`1<System.String>, System.String)' to access field 'System.Collections.Generic.List`1<System.__Canon>._size' failed.
Stack Trace:
    at dynamic_System.Collections.Generic.List`1[System.String]_Add(List`1 , String )
--- End of inner exception stack trace ---
    at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
   at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
   at System.Delegate.DynamicInvokeImpl(Object[] args)
   at System.Delegate.DynamicInvoke(Object[] args)
   at Pose.Tests.MethodRewriterTests.TestInstanceMethodRewrite() in C:\Code\tonerdo\pose\test\Pose.Tests\IL\MethodRewriterTests.cs:line 36


Total tests: 27. Passed: 22. Failed: 5. Skipped: 0.
Test Run Failed.
Test execution time: 1.1821 Seconds
```

```
.NET Command Line Tools (2.0.2)

Product Information:
 Version:            2.0.2
 Commit SHA-1 hash:  a04b4bf512

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.15063
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\2.0.2\

Microsoft .NET Core Shared Framework Host

  Version  : 2.0.0
  Build    : e8b8861ac7faf042c87a5c2f9f2d04c98b69f28d
```

--- a/Pose.sln
+++ b/Pose.sln
@@ -1,15 +1,14 @@
-
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
-VisualStudioVersion = 15.0.26124.0
+VisualStudioVersion = 15.0.27004.2002
 MinimumVisualStudioVersion = 15.0.26124.0
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "src", "src", "{3D859125-47BB-49F0-ACB4-42CEE47FB562}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Pose", "src\Pose\Pose.csproj", "{BBC1C0DC-ECA2-48C3-A233-0873428203AF}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Pose", "src\Pose\Pose.csproj", "{BBC1C0DC-ECA2-48C3-A233-0873428203AF}"
 EndProject
 Project("{2150E333-8FDC-42A3-9474-1A3956D46DE8}") = "test", "test", "{C2484D4C-4BED-46C2-BD61-2508BD8BD3C5}"
 EndProject
-Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Pose.Tests", "test\Pose.Tests\Pose.Tests.csproj", "{AE06C68C-032F-45CC-888C-54B0B98F860C}"
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Pose.Tests", "test\Pose.Tests\Pose.Tests.csproj", "{AE06C68C-032F-45CC-888C-54B0B98F860C}"
 EndProject
 Global
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
@@ -20,37 +19,40 @@ Global
 		Release|x64 = Release|x64
 		Release|x86 = Release|x86
 	EndGlobalSection
-	GlobalSection(SolutionProperties) = preSolution
-		HideSolutionNode = FALSE
-	EndGlobalSection
 	GlobalSection(ProjectConfigurationPlatforms) = postSolution
 		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x64.ActiveCfg = Debug|x64
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x64.Build.0 = Debug|x64
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x86.ActiveCfg = Debug|x86
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x86.Build.0 = Debug|x86
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x64.Build.0 = Debug|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Debug|x86.Build.0 = Debug|Any CPU
 		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|Any CPU.Build.0 = Release|Any CPU
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x64.ActiveCfg = Release|x64
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x64.Build.0 = Release|x64
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x86.ActiveCfg = Release|x86
-		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x86.Build.0 = Release|x86
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x64.ActiveCfg = Release|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x64.Build.0 = Release|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x86.ActiveCfg = Release|Any CPU
+		{BBC1C0DC-ECA2-48C3-A233-0873428203AF}.Release|x86.Build.0 = Release|Any CPU
 		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|Any CPU.ActiveCfg = Debug|Any CPU
 		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|Any CPU.Build.0 = Debug|Any CPU
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x64.ActiveCfg = Debug|x64
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x64.Build.0 = Debug|x64
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x86.ActiveCfg = Debug|x86
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x86.Build.0 = Debug|x86
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x64.ActiveCfg = Debug|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x64.Build.0 = Debug|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x86.ActiveCfg = Debug|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Debug|x86.Build.0 = Debug|Any CPU
 		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|Any CPU.ActiveCfg = Release|Any CPU
 		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|Any CPU.Build.0 = Release|Any CPU
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x64.ActiveCfg = Release|x64
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x64.Build.0 = Release|x64
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x86.ActiveCfg = Release|x86
-		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x86.Build.0 = Release|x86
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x64.ActiveCfg = Release|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x64.Build.0 = Release|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x86.ActiveCfg = Release|Any CPU
+		{AE06C68C-032F-45CC-888C-54B0B98F860C}.Release|x86.Build.0 = Release|Any CPU
+	EndGlobalSection
+	GlobalSection(SolutionProperties) = preSolution
+		HideSolutionNode = FALSE
 	EndGlobalSection
 	GlobalSection(NestedProjects) = preSolution
 		{BBC1C0DC-ECA2-48C3-A233-0873428203AF} = {3D859125-47BB-49F0-ACB4-42CEE47FB562}
 		{AE06C68C-032F-45CC-888C-54B0B98F860C} = {C2484D4C-4BED-46C2-BD61-2508BD8BD3C5}
+	EndGlobalSection
+	GlobalSection(ExtensibilityGlobals) = postSolution
+		SolutionGuid = {FE2B0751-A19A-4621-A703-18607E69D31A}
 	EndGlobalSection
 EndGlobal

--- a/src/Pose/Helpers/StubHelper.cs
+++ b/src/Pose/Helpers/StubHelper.cs
@@ -47,7 +47,6 @@ namespace Pose.Helpers
             return type.GetMethod(methodInfo.Name, bindingFlags, null, types, null);
         }
 
-        public static Module GetOwningModule()
-            => Assembly.GetAssembly(typeof(StubHelper)).Modules.FirstOrDefault();
+        public static Module GetOwningModule() => typeof(StubHelper).Module;
     }
 }

--- a/src/Pose/Helpers/StubHelper.cs
+++ b/src/Pose/Helpers/StubHelper.cs
@@ -46,5 +46,8 @@ namespace Pose.Helpers
             Type[] types = methodInfo.GetParameters().Select(p => p.ParameterType).ToArray();
             return type.GetMethod(methodInfo.Name, bindingFlags, null, types, null);
         }
+
+        public static Module GetOwningModule()
+            => Assembly.GetAssembly(typeof(StubHelper)).Modules.FirstOrDefault();
     }
 }

--- a/src/Pose/IL/MethodRewriter.cs
+++ b/src/Pose/IL/MethodRewriter.cs
@@ -38,7 +38,9 @@ namespace Pose.IL
             DynamicMethod dynamicMethod = new DynamicMethod(
                 string.Format("dynamic_{0}_{1}", _method.DeclaringType, _method.Name),
                 returnType,
-                parameterTypes.ToArray());
+                parameterTypes.ToArray(),
+                StubHelper.GetOwningModule(),
+                true);
 
             MethodDisassembler disassembler = new MethodDisassembler(_method);
             IList<LocalVariableInfo> locals = _method.GetMethodBody().LocalVariables;

--- a/src/Pose/IL/Stubs.cs
+++ b/src/Pose/IL/Stubs.cs
@@ -34,7 +34,9 @@ namespace Pose.IL
             DynamicMethod stub = new DynamicMethod(
                 string.Format("stub_{0}_{1}", methodInfo.DeclaringType, methodInfo.Name),
                 methodInfo.ReturnType,
-                parameterTypes.ToArray());
+                parameterTypes.ToArray(),
+                StubHelper.GetOwningModule(),
+                true);
 
             ILGenerator ilGenerator = stub.GetILGenerator();
 
@@ -108,7 +110,9 @@ namespace Pose.IL
             DynamicMethod stub = new DynamicMethod(
                 string.Format("stub_{0}_{1}", methodInfo.DeclaringType, methodInfo.Name),
                 methodInfo.ReturnType,
-                parameterTypes.ToArray());
+                parameterTypes.ToArray(),
+                StubHelper.GetOwningModule(),
+                true);
 
             ILGenerator ilGenerator = stub.GetILGenerator();
 
@@ -192,7 +196,9 @@ namespace Pose.IL
             DynamicMethod stub = new DynamicMethod(
                 string.Format("stub_{0}_{1}", constructorInfo.DeclaringType, constructorInfo.Name),
                 opCode == OpCodes.Newobj ? constructorInfo.DeclaringType : typeof(void),
-                parameterTypes.ToArray());
+                parameterTypes.ToArray(),
+                StubHelper.GetOwningModule(),
+                true);
 
             ILGenerator ilGenerator = stub.GetILGenerator();
             ilGenerator.DeclareLocal(constructorInfo.DeclaringType);
@@ -253,7 +259,9 @@ namespace Pose.IL
             DynamicMethod stub = new DynamicMethod(
                 string.Format("stub_{0}_{1}", methodInfo.DeclaringType, methodInfo.Name),
                 typeof(IntPtr),
-                parameterTypes.ToArray());
+                parameterTypes.ToArray(),
+                StubHelper.GetOwningModule(),
+                true);
 
             ILGenerator ilGenerator = stub.GetILGenerator();
             ilGenerator.DeclareLocal(typeof(MethodInfo));

--- a/test/Pose.Tests/Helpers/StubHelperTests.cs
+++ b/test/Pose.Tests/Helpers/StubHelperTests.cs
@@ -73,5 +73,12 @@ namespace Pose.Tests
             MethodInfo methodInfo = type.GetMethod("TestGetRuntimeMethodForVirtual");
             Assert.AreEqual(methodInfo, StubHelper.GetRuntimeMethodForVirtual(type, methodInfo));
         }
+
+        [TestMethod]
+        public void TestGetOwningModule()
+        {
+            Assert.AreEqual(typeof(StubHelper).Module, StubHelper.GetOwningModule());
+            Assert.AreNotEqual(typeof(StubHelperTests).Module, StubHelper.GetOwningModule());
+        }
     }
 }

